### PR TITLE
1858: Make it possible to configure to require reviewers for "Merge" PRs

### PR DESCRIPTION
**Pull request by zhaosongzs**
In this patch, PR bot has been added with the capability to control the requirement of reviews for merge pull requests.

By default, review is not necessary for merge pull requests.

However, if it is desired to enable the requirement of reviews for merge requests in a repo, the configuration can be added to the PR bot configuration as follows:
```
{
  "pr": {
      "repositories": {
          "repo1": {
               "reviewMerge": true
          }
       }
  }
}
```

Currently, jcheck would run in two places in skara bot. One is in checkRun and one is in IntegrateCommand. For the merge pull requests that not require review, the Reviewers check would be skipped.

The other way to run jcheck in using SKARA CLI. This patch is also compatible with it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1858](https://bugs.openjdk.org/browse/SKARA-1858): Make it possible to configure to require reviewers for "Merge" PRs

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1495/head:pull/1495` \
`$ git checkout pull/1495`

Update a local copy of the PR: \
`$ git checkout pull/1495` \
`$ git pull https://git.openjdk.org/skara.git pull/1495/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1495`

View PR using the GUI difftool: \
`$ git pr show -t 1495`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1495.diff">https://git.openjdk.org/skara/pull/1495.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1495#issuecomment-1492457986)

### CODE REVIEW

File: bots/pr/src/test/java/org/openjdk/skara/bots/pr/PullRequestBotFactoryTest.java [hunk 1]
```diff
@@ -112,7 +112,8 @@ public void testCreate() {
                             "integrator1",
                             "integrator2"
                           ],
-                          "reviewCleanBackport": true
+                          "reviewCleanBackport": true,
+                          "reviewMergePullRequest": true,
```
Comment (erikj79) [inline bots/pr/src/test/java/org/openjdk/skara/bots/pr/PullRequestBotFactoryTest.java:116]: I think `reviewMerge` is enough for the configuration option. The word "PullRequest" seems redundant in the context of pr bot.

File: jcheck/src/main/java/org/openjdk/skara/jcheck/ReviewersCheck.java [hunk 2]
```diff
@@ -67,7 +67,7 @@ private String nextRequiredRole(String role, Map<String, Integer> count) {
 
     @Override
     Iterator<Issue> check(Commit commit, CommitMessage message, JCheckConfiguration conf, Census census) {
-        if (commit.isMerge() || utils.addsHgTag(commit)) {
+        if (utils.addsHgTag(commit)) {
```
Comment (erikj79) [inline jcheck/src/main/java/org/openjdk/skara/jcheck/ReviewersCheck.java:70]: We can't just remove this default behavior in jcheck itself. Then the cli version of jcheck would always require reviewers for a merge commit. We are going to need a new setting in `.jcheck/conf` for this and the PR bot would override it when creating the jcheck instance.

Comment (zhaosongzs) [reply jcheck/src/main/java/org/openjdk/skara/jcheck/ReviewersCheck.java:70]: Ok, got it. If we have to make changes to `.jcheck/conf`. I will rework on this issue. It would be much more easy than current solution.

Comment (erikj79) [reply jcheck/src/main/java/org/openjdk/skara/jcheck/ReviewersCheck.java]: I think you misunderstood me. It needs to be possible to configure this from `.jcheck/conf`, but we still want to also be able to control it in the PR bot configuration for a specific repository. In the latter case it would be inserted as a config override when create the jcheck instance.

We want to enable this for the jdk main repository, but it should not get inherited by all downstream and project repos.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java [hunk 3]
```diff
@@ -116,7 +116,7 @@ public List<Bot> create(BotConfiguration configuration) {
             var censusRef = configuration.repositoryRef(repo.value().get("census").asString());
             var repository = configuration.repository(repo.name());
             var botBuilder = PullRequestBot.newBuilder()
-                                           .repo(repository)
+                                           .repo(configuration.repository(repo.name()))
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java:119]: Why this change? The variable repository is defined on the line above.

File: bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java [hunk 4]
```diff
@@ -127,6 +123,100 @@ void branchMerge(TestInfo testInfo) throws IOException {
...
+                commits = tempCommits.stream()
+                        .map(Commit::hash)
+                        .collect(Collectors.toSet());
+            }
+            assertTrue(commits.contains(otherHash1));
+            assertTrue(commits.contains(otherHash2));
+            assertFalse(commits.contains(mergeHash));
+
+            // Author and committer should updated in the merge commit
+            var headCommit = pushedRepo.commits(headHash.hex() + "^.." + headHash.hex()).asList().get(0);
```
Comment (erikj79) [inline bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java:211]: Does the commit message contain a "Reviewed-by" line?

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java (hunk 3 above)
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/PullRequestBotFactory.java:119]: When merging SKARA-1850 to this one, I must made a mistake. Thanks for catching it!

File: bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java (hunk 4 above)
Comment (zhaosongzs) [reply bots/pr/src/test/java/org/openjdk/skara/bots/pr/MergeTests.java:211]: Yes, it should have. I will update the test. Thx

**Review by erikj79 — APPROVED** [review]
